internal/api/controllers/private/highlevelConnectionStatus.go

```diff
@@ -0,0 +1,240 @@
+}
+
+func sortHostsByRecipient(details []inventory.HostDetails) (satelliteDetails []inventory.HostDetails, directConnectedDetails []inventory.HostDetails, noRhc []inventory.HostDetails) {
+	recipients := make(map[int][]inventory.HostDetails, 3)
+
+	for _, host := range details {
+		switch {
+		case host.SatelliteInstanceID != nil:
+			recipients[0] = append(recipients[0], host) // If satellite_instance_id exitsts Satellite host
+		case host.RHCClientID != nil:
+			recipients[1] = append(recipients[1], host) // if rhc_client_id exists in inventory facts host is direct connect
+		default:
+			recipients[2] = append(recipients[2], host)
+		}
+	}
+
+	return recipients[0], recipients[1], recipients[2]
+}
+
+func formatConnectionResponse(satID *string, satOrgID *string, rhcClientID *string, orgID OrgId, hosts []string, recipientType string, status string) RecipientWithConnectionInfo {
+	formatedHosts := make([]HostId, len(hosts))
+	var formatedSatID SatelliteId
+	var formatedSatOrgID SatelliteOrgId
+	var formatedRHCClientID public.RunRecipient
+
+	if satID != nil {
+		formatedSatID = SatelliteId(*satID)
+	}
+
+	if satOrgID != nil {
+		formatedSatOrgID = SatelliteOrgId(*satOrgID)
+	}
+
+	if rhcClientID != nil {
+		formatedRHCClientID = public.RunRecipient(*rhcClientID)
+	}
+
+	for i, host := range hosts {
+		formatedHosts[i] = HostId(host)
+	}
+
+	connectionInfo := RecipientWithConnectionInfo{
+		OrgId:         orgID,
+		Recipient:     formatedRHCClientID,
+		RecipientType: RecipientType(recipientType),
+		SatId:         formatedSatID,
+		SatOrgId:      formatedSatOrgID,
+		Status:        status,
+		Systems:       formatedHosts,
+	}
+
+	return connectionInfo
+}
+
+func getDirectConnectStatus(ctx echo.Context, client connectors.CloudConnectorClient, orgId OrgId, hostDetails []inventory.HostDetails) ([]RecipientWithConnectionInfo, error) {
+	responses := []RecipientWithConnectionInfo{}
+	for _, host := range hostDetails {
+		status, err := client.GetConnectionStatus(ctx.Request().Context(), string(orgId), *host.RHCClientID)
+
+		if err != nil {
+			utils.GetLogFromEcho(ctx).Error(err)
+			return nil, ctx.NoContent(http.StatusInternalServerError)
+		}
+
+		recipient := RecipientWithOrg{
+			OrgId:     orgId,
+			Recipient: public.RunRecipient(*host.RHCClientID),
+		}
+
+		results := recipientStatusResponse(recipient, status == connectors.ConnectionStatus_connected)
+
+		var connectionStatus string
+		if results.Connected {
+			connectionStatus = "connected"
+		} else {
+			connectionStatus = "disconnected"
+		}
+
+		responses = append(responses, formatConnectionResponse(nil, nil, host.RHCClientID, orgId, []string{host.ID}, string(RecipientType_directConnect), connectionStatus))
+	}
+
+	return responses, nil
+}
+
+func getSatelliteStatus(ctx echo.Context, client connectors.CloudConnectorClient, sourceClient sources.SourcesConnector, orgId OrgId, hostDetails []inventory.HostDetails) ([]RecipientWithConnectionInfo, error) {
```

Comment:
This function includes a couple of different sections that are delimited by comments.  You might consider factoring those out into functions themselves.  It might make the overall algorithm easier to follow.